libraries/chain/host_context.cpp

```diff
@@ -39,54 +39,83 @@ host_context::~host_context() = default;
 
 // called from apply_context or sync_call_context
 int64_t host_context::execute_sync_call(name call_receiver, uint64_t flags, std::span<const char> data) {
-   const uint32_t depth = sync_call_depth + 1;
-   const auto max_depth = control.get_global_properties().configuration.max_sync_call_depth;
-   EOS_ASSERT(depth <= max_depth, sync_call_depth_exception,
-              "reached sync call max call depth ${max_depth}", ("max_depth", max_depth));
+   auto start = fc::time_point::now();
 
-   const auto max_data_size = control.get_global_properties().configuration.max_sync_call_data_size;
-   EOS_ASSERT(data.size() <= max_data_size, sync_call_call_data_exception,
-              "sync call call data size must be less or equal to ${max_data_size} bytes", ("max_data_size", max_data_size));
+   // As early as possible, create the call trace of this new sync call in the parent's
+   // (sender's) trace to record entire trace of the sync call, including any exceptions
+   auto& trace = get_root_action_trace();
+   const bool read_only = flags & static_cast<uint64_t>(sync_call_flags::read_only);
+   trace.call_traces.emplace_back(get_sync_call_ordinal(), call_receiver, read_only, data);
 
-   const auto* code = control.db().find<account_object, by_name>(call_receiver);
-   EOS_ASSERT(code != nullptr, sync_call_validate_exception,
-              "sync call's receiver account ${r} does not exist", ("r", call_receiver));
-
-   EOS_ASSERT(flags <= static_cast<uint64_t>(sync_call_flags::all_allowed_bits), sync_call_validate_exception,  // all but `std::bit_width(all_allowed_bits)` LSBs must be 0s
-              "only ${bits} least significant bits of sync call's flags (${flags}) can be set", ("bits", std::bit_width(static_cast<uint64_t>(sync_call_flags::all_allowed_bits)))("flags", flags));
+   uint32_t ordinal = trace.call_traces.size();
+   get_call_trace(ordinal).call_ordinal = ordinal;
 
    auto handle_exception = [&](const auto& e)
    {
-      // TBD need to store the exception in the sync call trace
+      auto& call_trace = get_call_trace(ordinal);
+      call_trace.error_code = controller::convert_exception_to_error_code(e);
+      call_trace.except = e;
+      finalize_call_trace(call_trace, start);
       throw;
    };
 
+   auto handle_call_failure = [&]()
+   {
+      auto& call_trace = get_call_trace(ordinal);
+      call_trace.error_id = -1;
+      finalize_call_trace(call_trace, start);
+      trx_context.checktime();
```

Comment:
Probably not strictly necessary -- since no new WASM was run in this case, the expiration handlers weren't touched and we didn't run a risk of missing an expiration. Fine to keep as is since we may need to take more of a look on how expiration works in the future